akka-actor/src/main/scala/akka/actor/dungeon/ReceiveTimeout.scala

```diff
@@ -12,56 +12,65 @@ import scala.concurrent.duration.Duration
 import scala.concurrent.duration.FiniteDuration
 
 private[akka] object ReceiveTimeout {
-  final val emptyReceiveTimeoutData: (Duration, Cancellable) = (Duration.Undefined, ActorCell.emptyCancellable)
+  import ActorCell._
+
+  private final val empty: ReceiveTimeoutValue =
+    ReceiveTimeoutValue(Duration.Undefined, emptyCancellable, Duration.Undefined)
+
+  final case class ReceiveTimeoutValue(current: Duration, task: Cancellable, previous: Duration)
+
+  object ReceiveTimeoutValue {
+    def apply(): ReceiveTimeoutValue = empty
+  }
 }
 
 private[akka] trait ReceiveTimeout { this: ActorCell =>
 
   import ReceiveTimeout._
   import ActorCell._
 
-  private var receiveTimeoutData: (Duration, Cancellable) = emptyReceiveTimeoutData
+  private var value: ReceiveTimeoutValue = ReceiveTimeoutValue()
 
-  final def receiveTimeout: Duration = receiveTimeoutData._1
+  final def receiveTimeout: Duration = value.current
 
-  final def setReceiveTimeout(timeout: Duration): Unit = receiveTimeoutData = receiveTimeoutData.copy(_1 = timeout)
+  final def setReceiveTimeout(timeout: Duration): Unit = {
+    val prev = value.current
+    value = value.copy(current = timeout, previous = prev)
+  }
 
   protected def checkReceiveTimeoutIfNeeded(message: Any): Unit =
     if (hasTimeoutData)
```

Comment:
I think this still misses if a timeout was removed during a `NotInfluenceReceiveTimeout` message handling.

Reply:
This comment is still valid, because if it's a NotInfluenceReceiveTimeout message it will not be cancelled.
Like this:
```
  val receiveTimeoutChanged = receiveTimeoutData ne beforeReceive
  if (hasTimeoutData || receiveTimeoutChanged)
    checkReceiveTimeout(!message.isInstanceOf[NotInfluenceReceiveTimeout] || receiveTimeoutChanged)
```

Reply:
This one really schooled me :) Thanks for the input, pushed change.